worker/upload.go

```diff
@@ -0,0 +1,1154 @@
+	span.SetAttributes(attribute.Stringer("hk", q.hk))
+	span.AddEvent("enqueued")
+	j.queue = q
+
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	// enqueue the job
+	q.queue = append(q.queue, j)
+
+	// signal there's work
+	select {
+	case q.queueChan <- struct{}{}:
+	default:
+	}
+}
+
+func (q *uploadQueue) track(err error, d time.Duration) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	if err != nil {
+		q.consecutiveFailures++
+		q.statsSpeed.track(math.SmallestNonzeroFloat64)
+		fmt.Println("TRACK ERROR", q.hk, err)
+	} else {
+		q.consecutiveFailures = 0
+		q.statsSpeed.track(mbps(rhpv2.SectorSize, d.Seconds()))
+	}
+}
+
+func (q *uploadQueue) updateBlockHeight(bh uint64) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	q.bh = bh
+}
+
+func (q *uploadQueue) pop() *uploadJob {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	if len(q.queue) > 0 {
+		j := q.queue[0]
+		q.queue = q.queue[1:]
+		return j
+	}
+	return nil
+}
+
+func (j *uploadJob) execute(hp hostProvider, rev types.FileContractRevision) (types.Hash256, error) {
+	// fetch span from context
+	span := trace.SpanFromContext(j.requestCtx)
+	span.AddEvent("execute")
+
+	// create a host
+	h, err := hp.newHostV3(j.requestCtx, j.queue.fcid, j.queue.hk, j.queue.siamuxAddr)
+	if err != nil {
+		return types.Hash256{}, err
+	}
+
+	// upload the sector
+	start := time.Now()
+	root, err := h.UploadSector(j.requestCtx, j.sector, rev)
+	if err != nil {
+		return types.Hash256{}, err
+	}
+
+	// update span
+	elapsed := time.Since(start)
+	span.SetAttributes(attribute.Int64("duration", elapsed.Milliseconds()))
+	span.RecordError(err)
+	span.End()
+	return root, nil
+}
+
+func (j *uploadJob) succeed(root types.Hash256) {
+	select {
+	case <-j.requestCtx.Done():
+	case j.responseChan <- sectorResponse{
+		job:  j,
+		root: root,
+	}:
+	}
+}
+
+func (j *uploadJob) fail(err error) {
+	select {
+	case <-j.requestCtx.Done():
+	case j.responseChan <- sectorResponse{
+		job: j,
+		err: err,
+	}:
+	}
+}
+
+func (j *uploadJob) done() bool {
+	select {
+	case <-j.requestCtx.Done():
+		return true
+	default:
+		return false
+	}
+}
+
+func (s *uploadState) launch(job *uploadJob) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// enqueue the job
+	err := s.u.enqueue(job)
+	if err != nil {
+		span := trace.SpanFromContext(job.requestCtx)
+		span.RecordError(err)
+		span.End()
+		return err
+	}
+
+	// update the state
+	s.numInflight++
+	s.numLaunched++
+	if job.overdrive {
+		s.lastOverdrive = time.Now()
+		s.overdriving[job.sectorIndex]++
+	}
+
+	return nil
+}
+
+func (s *uploadState) receive(resp sectorResponse) (completed bool) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// failed jobs can't complete the upload
+	s.numInflight--
+	if resp.err != nil {
+		s.errs = append(s.errs, &HostError{resp.job.queue.hk, resp.err})
+		return false
+	}
+
+	defer func() {
+		s.u.registerCompletedSector(s.uploadID, s.shardID, resp.job.queue.fcid, completed)
+	}()
+
+	// redundant sectors can't complete the upload
+	s.numCompleted++
```

Comment:
Is this only used for computing the performance? If so, should it really be incremented for redundant sectors? Imo this will give a wrong impression about the actual performance since more overdrive will make it seem like the upload was faster than it actually was.